### ISSUE TITLE
Potential incorrect results when scanning many parquet files

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#6125]
I thought of this problem while working on teaching DataFusion about how the parquet files are sorted in IOx: https://github.com/influxdata/influxdb_iox/pull/6108. I don't think that PR makes this problem any better / worse so I have filed a separate ticket

# Background

The [iox provider](https://github.com/influxdata/influxdb_iox/blob/main/iox_query/src/provider.rs#L511) makes physical plans like this:

```text
                                     ┌───────────────────────┐
                                     │SortPreservingMergeExec│   <-- This is added if sort_output = true
                                     └───────────────────────┘
                                              ▲
                                              │
                                     ┌─────────────────┐
                                     │    UnionExec    │
                                     │                 │
                                     └─────────────────┘
                                              ▲
                                              │
                       ┌──────────────────────┴───────────┬─────────────────────┐
                       │                                  │                     │
                       │                                  │                     │
              ┌─────────────────┐                ┌─────────────────┐   ┌─────────────────┐
              │ DeduplicateExec │                │ DeduplicateExec │   │     SortExec    │  <-- This is added if output_sort_key.is_some()
              └─────────────────┘                └─────────────────┘   │    (Optional)   │
                       ▲                                  ▲            └─────────────────┘
                       │                                  │                     ▲
           ┌───────────────────────┐                      │                     │
           │SortPreservingMergeExec│                      │             ┌─────────────────┐
           └───────────────────────┘                      │             │   ParquetExec   │
                       ▲                                  │             │    (Chunk 4)    │
                       │                                  │             └─────────────────┘
           ┌───────────────────────┐                      │
           │       UnionExec       │                      │
           └───────────────────────┘                      │
                      ▲                                   |
                      │                                   |
          ┌───────────┴───────────┐                       │
          │                       │                       │
 ┌─────────────────┐     ┌─────────────────┐    ┌─────────────────┐
 │    SortExec     │     │    SortExec     │    │    SortExec     │
 │   (optional)    │     │   (optional)    │    │   (optional)    │
 └─────────────────┘     └─────────────────┘    └─────────────────┘
          ▲                       ▲                      ▲
          │                       │                      │
          │                       │                      │
 ┌─────────────────┐     ┌─────────────────┐    ┌─────────────────┐
 │   ParquetExec   │     │   ParquetExec   │    │   ParquetExec   │
 │    (Chunk 1)    │     │    (Chunk 2)    │    │    (Chunk 3)    │
 └─────────────────┘     └─────────────────┘    └─────────────────┘
```

There is an optional `SortExec` in all the branches that is omitted if the data is already sorted correctly

# Problem

However, after https://github.com/influxdata/influxdb_iox/pull/6072, IOx uses  DataFusion's `ParquetExec` reader directly, I think sometimes these sorts will be incorrectly omitted resulting in potentially lost data (as streams will be merged that are not sorted).

To understand why, it is important to know that one of the reasons we use the ParuqetExec is that it will not attempt to read all files concurrently instead it will read files in sequence within a **DataFusion Partition** (as opposed to an IOx partition).

This is an important property as it allows plans to run against 1000s of parquet files and not try to open them all concurrently.

However, it means if we assign more than one parquet file to a partitition the output sort order will not be preserved as illustrated in this diagrams

When only 1 file is assigned to each partition, each partition is correctly sorted on `(A, B, C)`

```text
┏ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ┓
  ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐ ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─  ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─  ┌ ─ ─ ─ ─ ─ ─ ─ ─ ┐
┃   ┌───────────────┐     ┌──────────────┐ │   ┌──────────────┐ │   ┌─────────────┐   ┃
  │ │   1.parquet   │ │ │ │  2.parquet   │   │ │  3.parquet   │   │ │  4.parquet  │ │
┃   │ Sort: A, B, C │     │Sort: A, B, C │ │   │Sort: A, B, C │ │   │Sort: A, B, C│   ┃
  │ └───────────────┘ │ │ └──────────────┘   │ └──────────────┘   │ └─────────────┘ │
┃                                          │                    │                     ┃
  │                   │ │                    │                    │                 │
┃                                          │                    │                     ┃
  │                   │ │                    │                    │                 │
┃                                          │                    │                     ┃
  │                   │ │                    │                    │                 │
┃  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘  ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘  ─ ─ ─ ─ ─ ─ ─ ─ ─  ┃
     DataFusion           DataFusion           DataFusion           DataFusion
┃    Partition 1          Partition 2          Partition 3          Partition 4       ┃
 ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━

                                      ParquetExec
```

When more than 1 file is assigned to each partition, each partition is NOT correctly sorted on `(A, B, C)`. Once the second file is scanned, the same values for A, B and C can be repeated in the same sorted stream

```text
┏ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━
  ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─ ┐ ┌ ─ ─ ─ ─ ─ ─ ─ ─ ─  ┃
┃   ┌───────────────┐     ┌──────────────┐ │
  │ │   1.parquet   │ │ │ │  2.parquet   │   ┃
┃   │ Sort: A, B, C │     │Sort: A, B, C │ │
  │ └───────────────┘ │ │ └──────────────┘   ┃
┃   ┌───────────────┐     ┌──────────────┐ │
  │ │   3.parquet   │ │ │ │  4.parquet   │   ┃
┃   │ Sort: A, B, C │     │Sort: A, B, C │ │
  │ └───────────────┘ │ │ └──────────────┘   ┃
┃                                          │
  │                   │ │                    ┃
┃  ─ ─ ─ ─ ─ ─ ─ ─ ─ ─   ─ ─ ─ ─ ─ ─ ─ ─ ─ ┘
     DataFusion           DataFusion         ┃
┃    Partition 1          Partition 2
 ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ━ ┛

              ParquetExec
```

# Impact analysis

I don't know how often this will happen (as the compactor does a pretty good job of keeping the number of overlapping parquet files down but I think we need to fix it as soon as possible

**Comment by alamb** [comment]
@crepererum  maybe we can discuss this on Monday

**Comment by alamb** [comment]
I spent all morning trying to coax a bug out and I was not able to. After consideration I believe the way the IOx query plans are constructed this will not actually cause issues 😅 

Eventually after consultation with @NGA-TRAN  I think we have convinced ourselves that the issue will not be a problem in practice

The core reason is that chunks_to_physical_nodes which makes parquet_exec is only called in two places:
1: `build_sort_plan_for_read_filter` [link](https://github.com/influxdata/influxdb_iox/blob/d0b3ac1cef12764c27b9328b8d241bde6ff5cf5c/iox_query/src/provider.rs#L1098) where it has exactly one chunk and thus is not affected
2. build_plans_for_non_duplicates_chunks [link](https://github.com/influxdata/influxdb_iox/blob/d0b3ac1cef12764c27b9328b8d241bde6ff5cf5c/iox_query/src/provider.rs#L1272-L1282)  will only use a single ParquetExecp` if the output does not need to be sorted. This is the case for queries but not for reorg plans (like the compactor)

So for example I had a test with :


`Chunk1`, `Chunk2` that maybe overlap and `Chunk3` and `Chunk4` that are not overlapping. The query planner created this plan:


```
SortPreservingMergeExec: [tag1@1 ASC,time@2 ASC]
  UnionExec
    DeduplicateExec: [tag1@1 ASC,time@2 ASC]
      SortPreservingMergeExec: [tag1@1 ASC,time@2 ASC]
        UnionExec
          ParquetExec: limit=None, partitions=[Chunk1] <-- created by build_sort_plan_for_read_filter
          ParquetExec: limit=None, partitions=[Chunk2] <-- created by build_sort_plan_for_read_filter
    ParquetExec: limit=None, partitions=[Chunk3] <--- created by build_plans_for_non_duplicates_chunks
    ParquetExec: limit=None, partitions=[Chunk4] <--- created by build_plans_for_non_duplicates_chunks
```

This does imply that compact plans (that require the output to be sorted) are still going to use one ParquetExec instance per file, but I think that is sort of required for now (to run them through a sort preserving merge)

I plan to polish up a few of my tests from earlier today to try and cover any potential regression , but otherwise I will now sleep more soundly
